pkg/agent/agent.go

```diff
@@ -87,6 +92,23 @@ func (a *Agent) readPump() {
 		return a.conn.SetReadDeadline(time.Now().Add(pongWait))
 	})
 
+	a.conn.SetCloseHandler(func(err int, msg string) error {
```

Comment:
I'm not sure if that's the best way to do that - https://godoc.org/github.com/gorilla/websocket#Conn.SetCloseHandler. First of all, it's possible that we replace the default handler so we lose some functionality. Second thing is that I'm not sure if it's always called. Wondering if it's called if a peer just drops the connection, without sending close message..

Should we maybe just do it in the defer function, after unregistering from the broker so it's not send to the disconnecting peer?

Reply:
One more thing that maybe you can look at, I noticed that there's sometimes a delay between peer disconnecting and the corresponding log message on the server. I didn't look at it because it wasn't an issue, but it might be now because we might want to know that someone disconnected ASAP.

Reply:
I think it was because server discovered that someone disconnected only when the ping timeout or it tried to ping an disconnected peer, but would need to be verified.

Reply:
I noticed it also, my suspicion is that in writePump() method when you call return on line 118 it won't actually exit the for loop but will wait for a next tick and then encounter closed connection, print log that it cannot send it and finnaly exit the coroutine.